vllm/worker/model_runner_base.py

```diff
@@ -0,0 +1,185 @@
+
+    # An empty SamplingMetadata to signal that the worker should skip
+    # sampling.
+    sampling_metadata = SamplingMetadata(
+        seq_groups=None,
+        selected_token_indices=selected_token_indices,
+        categorized_sample_indices=None,
+        num_prompts=0,
+    )
+    kwargs["sampling_metadata"] = sampling_metadata
+    return kwargs
+
+
+def _add_sampling_metadata_broadcastable_dict(
+        tensor_dict: Dict[str, Union[int, torch.Tensor]],
+        sampling_metadata: Optional["SamplingMetadata"]) -> None:
+    """
+    Helper method to update tensor_dict with broadcastable
+    SamplingMetadata fields.
+    """
+    if sampling_metadata is not None:
+        tensor_dict["selected_token_indices"] = (
+            sampling_metadata.selected_token_indices)
+
+
+def _filter_valid_kwargs(cls, kwargs: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Helper method to filter the given kwargs to kwargs that
+    are valid for the given dataclass `cls`.
+    """
+    init_kwargs = {}
+    for field in dataclasses.fields(cls):
+        val = kwargs.get(field.name, None)
+        if val is not None:
+            init_kwargs[field.name] = val
+    return init_kwargs
+
+
+@dataclasses.dataclass(frozen=True)
+class ModelRunnerInputBase(ABC):
+    """Local inputs to each worker's model runner. May contain
+    device-specific data. Different worker backends may have different methods
+    of converting from the global ExecuteModelRequest produced by the LLM
+    engine to the worker-local ModelRunnerInputBase objects.
+
+    Model runners that support multi-GPU execution should define a
+    ModelRunnerInputBase subclass, add their required fields, and specify how to
+    serialize/deserialize a ModelInput for broadcast between workers.
+    """
+
+    @classmethod
+    @abstractmethod
+    def new(cls: Type[T], **kwargs) -> T:
+        """
+        Create a new instance of this class. Populate the new instance with
+        the given kwargs.
+        """
+        raise NotImplementedError
+
+    def replace(self: T, **kwargs) -> T:
+        """
+        Replace current fields with fields in kwargs.
+        """
+        return dataclasses.replace(self, **kwargs)
+
+    def as_broadcastable_tensor_dict(
+            self) -> Dict[str, Union[int, torch.Tensor]]:
+        """
+        Extract broadcastable fields. Override for fields that require some
+        custom deserialization.
+        """
+        raise NotImplementedError
+
+    def _get_attrs(self, attrs: List[str]) -> Dict[str, Any]:
+        """
+        Helper method to get a dictionary from attribute name to value.
+        Attributes whose values are None will not be added to the returned
+        dictionary.
+        """
+        tensor_dict: Dict[str, Union[int, torch.Tensor]] = {}
+        for attr in attrs:
+            val = getattr(self, attr, None)
+            if val is not None:
+                tensor_dict[attr] = val
+
+        return tensor_dict
```

Comment:
IMHO, if an attribute is `None`, we can also put it in the returning dict. We don't need to have a special logic for this.

Reply:
`broadcast_object ` can accept value of `None`. All non-tensor parts will be squeezed into a python object, without additional `broadcast` call.

Reply:
But there is still serialization cost, no?

Reply:
just a few more bytes in the serialized bytes, I don't think it is worth optimizing.